lib/ansible/modules/cloud/vmware/vmware_datastore_upload.py

```diff
@@ -0,0 +1,316 @@
+            vim.SessionManagerGenericServiceTicket
+        '''
+        return self.content.sessionManager.AcquireGenericServiceTicket(
+            spec=PyVmomi.vim.SessionManagerHttpServiceRequestSpec(
+                url=url, method='http' + method.title()))
+
+    def _destination_exists(self):
+        '''
+            Check to see if the file already exists and has a corresponding
+            .sha256 file on the destination.  If so, this operation can be
+            idempotent by reading the .sha256 file, and calculating the local
+            SHA256, and if they match then there is nothing to do.
+
+            Returns:
+                ( same, hash )
+        '''
+        local_hash = self._local_hash()
+        self.module.debug('LOCAL  SHA256: ' + local_hash)
+        remote_hash = self._remote_hash()
+        self.module.debug('REMOTE SHA256: ' + str(remote_hash))
+        return (local_hash == remote_hash, local_hash)
+
+    def _expand(self, url, params):
+        ''' Returns the complete url, needed to request a ticket. '''
+        return requests.Request('GET', url, params=params).prepare().url
+
+    def _get(self, url, headers=None, params=None):
+        ''' Issue a GET - on 404 response return is None otherwise text '''
+        # TODO: add certificate verification
+        r = requests.get(url, headers=headers, params=params, verify=False)
+        if r.status_code == 404:
+            return None
+        if r.status_code == 200:
+            return r.text
+        raise Exception('GET {0} => {1}'.format(url, r.status_code))
+
+    def _local_hash(self):
+        '''
+            Obtain and optionally cache the local hash.
+        '''
+        generate_sha256 = True
+        if self.debounce_cache_local:
+            # we regenerate the hash if the .sha256 file is
+            # older than the source
+            source_stat = os.stat(self.source_path)
+            if os.path.exists(self.source_sha256_path):
+                sha256_stat = os.stat(self.source_sha256_path)
+                generate_sha256 = source_stat.st_mtime >= sha256_stat.st_mtime
+        if generate_sha256:
+            hash = hashlib.sha256(self.data).hexdigest()
+            if self.debounce_cache_local:
+                with open(self.source_sha256_path, "w") as file:
+                    file.write(hash)
+            return hash
+        with open(self.source_sha256_path, "r") as file:
+            return file.read()
+
+    def _put(self, dest, data):
+        ''' Perform an upload. '''
+        url, params = self._vmware_datastore_io_url(dest)
+        ticket = self._acquire_service_ticket(self._expand(url, params), 'PUT')
+        headers = {
+            "Content-Type": "application/octet-stream",
+            "Content-Length": str(len(data)),
+            "Cookie": 'vmware_cgi_ticket=' + ticket.id
+        }
+        # TODO: add certificate verification
+        r = requests.put(url, headers=headers, params=params, data=data, verify=False)
```

Comment:
Make `verify` configurable via parameter.

Reply:
There's more to it than just that - one needs the ability to specify the client certificate, key, and CA if one is going to do certificate verification.

Reply:
https://github.com/ansible/ansible/blob/cd988f645aaf774c55a98a90e3ef42cc5b1a1563/lib/ansible/module_utils/urls.py#L1147 gives all options.